operators/src/upgrade/rest/service.rs

```diff
@@ -125,9 +138,179 @@ pub async fn apply_upgrade() -> Result<HttpResponse, RestError> {
+            }
+        }
+    }
+    Ok(())
+}
+
+pub async fn is_data_plane_pod_running(node: &str) -> Result<bool, Error> {
+    let mut data_plane_pod_running = false;
+    let pods: Api<Pod> = Api::namespaced(
+        UpgradeConfig::get_config().k8s_client().client(),
+        UpgradeConfig::get_config().namespace(),
+    );
+    let io_engine_listparam = ListParams::default().labels(IO_ENGINE_POD_LABEL);
+    let initial_io_engine_pod_list: ObjectList<Pod> = pods.list(&io_engine_listparam).await?;
+    //let data_plane_pod_running =
+    for pod in initial_io_engine_pod_list.iter() {
+        // Fetch the node name on which the io-engine pod is running
+        let node_name = pod
+            .spec
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("pod.spec.node_name is invalid"))?
+            .node_name
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("pod.spec.node_name is invalid"))?
+            .as_str();
+        if node != node_name {
+            continue;
+        } else {
+            match pod
+                .status
+                .as_ref()
+                .and_then(|status| status.conditions.as_ref())
+            {
+                Some(conditions) => {
+                    for condition in conditions {
+                        if condition.type_.eq("Ready") && condition.status.eq("True") {
+                            data_plane_pod_running = true
+                        } else {
+                            data_plane_pod_running = false;
+                        }
+                    }
+                }
+                None => {
+                    data_plane_pod_running = false;
+                }
+            }
+        }
+    }
+    Ok(data_plane_pod_running)
+}
+
+pub async fn is_control_plane_running() -> Result<(), Error> {
+    let pods: Api<Pod> = Api::namespaced(
+        UpgradeConfig::get_config().k8s_client().client(),
+        UpgradeConfig::get_config().namespace(),
+    );
+
+    let pod_list: ObjectList<Pod> = pods
+        .list(&ListParams::default().labels(AGENT_CORE_POD_LABEL))
+        .await?;
+    if !all_pods_are_ready(pod_list) {
+        return Err(Error::ControlPlanePodsDown(
+            "Helm not installed".to_string(),
+        ));
+    }
+
+    let pod_list: ObjectList<Pod> = pods
+        .list(&ListParams::default().labels(API_REST_POD_LABEL))
+        .await?;
+    if !all_pods_are_ready(pod_list) {
+        return Err(Error::ControlPlanePodsDown(
+            "Helm not installed".to_string(),
+        ));
+    }
+
+    let pod_list: ObjectList<Pod> = pods
+        .list(&ListParams::default().labels(ETCD_POD_LABEL))
+        .await?;
+    if !all_pods_are_ready(pod_list) {
+        return Err(Error::ControlPlanePodsDown(
+            "Helm not installed".to_string(),
+        ));
+    }
+    Ok(())
+}
+
 /// Get  upgrade.
 #[get("/upgrade")]
 pub async fn get_upgrade() -> impl Responder {
+    upgrade_data_plane().await;
```

Comment:
oh, we're upgrading the data plane on a get? Seems to violate http rules... gets should be safe operations